Typing fixes

michaelosthege opened the pull request:
This is a follow-up to #82 where Ricardo noticed that a `Tuple` type hint was missing `...`.

I searched the entire codebase for similar `Tuple`-related inaccuracies and found many.

This prompted me to change the way how `mypy` is invoked to the `run_mypy.py` approach we use in PyMC.
I directly used the inverted logic version from https://github.com/pymc-devs/pymc/pull/6381 to keep the list short.

This approach makes it easier to get an overview on unresolved mypy errors (`python scripts/run_mypy.py`) that were previously suppressed by the settings from `setup.cfg`.

ricardoV94 commented:
About the None vs Optional is there a difference? I tend to use None when that's a valid argument but not Optional (if you don't provide it errors), but I didn't follow any guidelines or anything.

For instance, in the SpecifyShape case it's not optional, but it's a valid type.

michaelosthege commented:
> About the None vs Optional is there a difference? I tend to use None when that's a valid argument but not Optional (if you don't provide it errors), but I didn't follow any guidelines or anything.
> 
> For instance, in the SpecifyShape case it's not optional, but it's a valid type.

[This](https://stackoverflow.com/a/51710151/4473230) is probably the most comprehensive answer, according to which they are identical.

I actually thought they weren't, which I why I changed it. I reverted that line because here the `Union[None, int, ...]` is actually more aesthetic.

[ready for review]

ricardoV94 reviewed (approved):
LGTM, let me know

maresb commented:
Sorry, I just saw this now, linked from [pymc-devs/pytensor#6381](https://github.com/pymc-devs/pymc/pull/6381). Best is to use the modern syntax e.g. `tuple[slice | Variable | None], ...]`. This works with older python versions simply by importing `__future__.annotations`. Indeed, just stick `from __future__ import annotations` up top and run `pyupgrade` via pre-commit and watch the magic. :wink:

maresb commented:
Adding a proper StackOverflow [answer](https://stackoverflow.com/a/74768735) (feedback appreciated)
